File: benchmarks/reference/merge_sort.go
package main
import ("fmt"; "time")
func merge(left, right []int) []int {
    var result []int
    i, j := 0, 0
    for i < len(left) && j < len(right) {
        if left[i] <= right[j] { result = append(result, left[i]); i++ } else { result = append(result, right[j]); j++ }
    }
    for i < len(left) { result = append(result, left[i]); i++ }
    for j < len(right) { result = append(result, right[j]); j++ }
    return result
}
func mergeSort(arr []int) []int {
    if len(arr) <= 1 { return arr }
    mid := len(arr) / 2
    left := mergeSort(arr[:mid])
    right := mergeSort(arr[mid:])
    return merge(left, right)
}
func main() {
    start := time.Now()
    var arr []int
    for i := 0; i < 5000; i++ { arr = append(arr, 5000 - i) }
    sorted := mergeSort(arr)
    fmt.Println("First:", sorted[0])
    fmt.Println("Last:", sorted[len(sorted)-1])
    fmt.Printf("Time: %.6f\n", time.Since(start).Seconds())
}

File: benchmarks/reference/merge_sort.cpp
#include <iostream>
#include <vector>
#include <chrono>

std::vector<int> merge(const std::vector<int>& left, const std::vector<int>& right) {
    std::vector<int> result;
    size_t i = 0, j = 0;
    while(i < left.size() && j < right.size()) {
        if(left[i] <= right[j]) result.push_back(left[i++]);
        else result.push_back(right[j++]);
    }
    while(i < left.size()) result.push_back(left[i++]);
    while(j < right.size()) result.push_back(right[j++]);
    return result;
}

std::vector<int> merge_sort(const std::vector<int>& arr) {
    if(arr.size() <= 1) return arr;
    size_t mid = arr.size() / 2;
    std::vector<int> left(arr.begin(), arr.begin() + mid);
    std::vector<int> right(arr.begin() + mid, arr.end());
    return merge(merge_sort(left), merge_sort(right));
}

int main() {
    auto start = std::chrono::high_resolution_clock::now();
    std::vector<int> arr;
    for(int i=0; i<5000; i++) arr.push_back(5000 - i);
    std::vector<int> sorted = merge_sort(arr);
    
    std::cout << "First: " << sorted[0] << "\n";
    std::cout << "Last: " << sorted.back() << "\n";
    std::chrono::duration<double> elapsed = std::chrono::high_resolution_clock::now() - start;
    std::cout << "Time: " << elapsed.count() << "\n";
    return 0;
}
